backend/mailru/mailru.go

```diff
@@ -0,0 +1,2356 @@
+	}
+	obj := &Object{
+		fs:          t.f,
+		remote:      remote,
+		hasMetaData: true,
+		size:        size,
+		mrHash:      binHash,
+		modTime:     modTime,
+	}
+	return obj, nil
+}
+
+type treeState struct {
+	f           *Fs
+	r           *api.BinReader
+	options     int
+	rootDir     string
+	currDir     string
+	lastDir     string
+	level       int
+	revision    treeRevision
+	totalSpace  int64
+	usedSpace   int64
+	fingerprint []byte
+	dunno1      int
+	dunno2      int
+	dunno3      int
+	dunnoDel1   int
+	dunnoDel2   int
+	dunnoNodeID []byte
+	treeID      []byte
+}
+
+type treeRevision struct {
+	ver       int16
+	treeID    []byte
+	treeIDNew []byte
+	bgn       uint64
+	bgnNew    uint64
+}
+
+func (rev *treeRevision) Read(data *api.BinReader) error {
+	rev.ver = data.ReadByteAsShort()
+	switch rev.ver {
+	case 0:
+		// Revision()
+	case 1, 2:
+		rev.treeID = data.ReadNBytes(api.TreeIDLength)
+		rev.bgn = data.ReadULong()
+	case 3, 4:
+		rev.treeID = data.ReadNBytes(api.TreeIDLength)
+		rev.bgn = data.ReadULong()
+		rev.treeIDNew = data.ReadNBytes(api.TreeIDLength)
+		rev.bgnNew = data.ReadULong()
+	case 5:
+		rev.treeID = data.ReadNBytes(api.TreeIDLength)
+		rev.bgn = data.ReadULong()
+		rev.treeIDNew = data.ReadNBytes(api.TreeIDLength)
+	default:
+		return fmt.Errorf("unknown directory revision %d", rev.ver)
+	}
+	return data.Error()
+}
+
+// CreateDir makes a directory (parent must exist)
+func (f *Fs) CreateDir(ctx context.Context, path string) error {
+	fs.Debugf(f, ">>> CreateDir %q", path)
+
+	req := api.NewBinWriter()
+	req.WritePu16(api.OperationCreateFolder)
+	req.WritePu16(0) // revision
+	req.WriteString(path)
+	req.WritePu32(0)
+
+	token, err := f.accessToken()
+	if err != nil {
+		return err
+	}
+	metaURL, err := f.metaServer()
+	if err != nil {
+		return err
+	}
+
+	opts := rest.Opts{
+		Method:  "POST",
+		RootURL: metaURL,
+		Parameters: url.Values{
+			"client_id": {api.OAuthClientID},
+			"token":     {token},
+		},
+		ContentType: api.BinContentType,
+		Body:        req.Reader(),
+	}
+
+	var res *http.Response
+	err = f.pacer.Call(func() (bool, error) {
+		res, err = f.srv.Call(&opts)
+		return shouldRetry(res, err, f, &opts)
+	})
+	if err != nil {
+		closeBody(res)
+		return err
+	}
+
+	reply := api.NewBinReader(res.Body)
+	defer closeBody(res)
+
+	switch status := reply.ReadByteAsInt(); status {
+	case api.MkdirResultOK:
+		return nil
+	case api.MkdirResultAlreadyExists, api.MkdirResultExistsDifferentCase:
+		return ErrorDirAlreadyExists
+	case api.MkdirResultSourceNotExists:
+		return ErrorDirSourceNotExists
+	case api.MkdirResultInvalidName:
+		return ErrorInvalidName
+	default:
+		return fmt.Errorf("mkdir error %d", status)
+	}
+}
+
+// Mkdir creates the container (and its parents) if it doesn't exist.
+// Normally it ignores the ErrorDirAlreadyExist, as required by rclone tests.
+// Nevertheless, such programs as borgbackup or restic use mkdir as a locking
```

Comment:
How does borgbackup call rclone? Maybe a flag to `rclone mkdir` might be better?

Reply:
Let me investigate and submit a separate Issue. The issue will state that the mailru backend quirk should be removed when the issue resolves.